### ISSUE TITLE
install-docker-ce.ps1 defaults to nightly builds

### ISSUE THREAD

**Issue by olljanat** [microsoft/Windows-Containers#256]
**Describe the bug**
`install-docker-ce.ps1` which is mentioned in [this](https://docs.microsoft.com/en-us/virtualization/windowscontainers/quick-start/set-up-environment?tabs=dockerce#windows-server-1) documentation defaults to non-released, nightly builds of Docker:
https://github.com/microsoft/Windows-Containers/blob/78519ddbc455916f677aead0aa91f5711c91f25d/helpful_tools/Install-DockerCE/install-docker-ce.ps1#L65-L71
which is very bad idea people will end up to seeing weird issues on their environments because of that. Proper way would be default to latest stable binaries which can be found from https://download.docker.com/win/static/stable/x86_64/

**To Reproduce**
Install like it is described in documentation.

**Expected behavior**
That script default to latest stable version of Docker.

**Configuration:**
N/A

**Additional context**
N/A

**Comment by fady-azmy-msft** [comment]
@brasmith-ms

**Comment by pawelpabich** [comment]
@fady-azmy-msft or @brasmith-ms can you comment on this issue re whether the current approach is not correct and should be fixed? Happy to submit PR unless there is a reason why installing the latest version from master makes sense.

**Comment by brasmith-ms** [comment]
Hey @olljanat @pawelpabich sorry must have missed this issue. This was meant for dev purposes but I do agree it's better to use the stable versions, just requires some version parsing logic. If you have a PR for this feel free to submit and approve, otherwise I'll be able to get to it over the next week.

**Comment by olljanat** [comment]
> just requires some version parsing logic.

Nowadays releases are created to https://github.com/moby/moby/releases so you should be able to borrow this logic to figure out what is latest stable version https://github.com/olljanat/alz-bicep-ci/blob/master/standard/scripts/update-alz-bicep.ps1#L2-L3

**Comment by pawelpabich** [comment]
Thanks @brasmith-ms for your clarifications.
I had a look and decided not to submit a PR mostly because this script is more than likely already used in the wild :). This means that some thought needs to be given to versioning. I don't feel I have enough information to make a good decision here.

I like @olljanat approach, but it assumes that both places get updated at more or less the same time.  Parsing HTML is the last thing we want to do, but that might be the most reliable approach here

```
$request = Invoke-WebRequest -Uri https://download.docker.com/win/static/stable/x86_64/ -UseBasicParsing
[regex]::matches($request, "docker\-\d+\.\d+\.\d+\.zip") | Foreach { Write-Host $_.Value }
```

Another approach might be to retrieve the list of releases from Moby and then download the most recent one available on the download page. 

A few decisions to make :).

**Comment by AJDurant** [comment]
These nightly images also have not been updated since 2022-03-27 which leads to issues with other programs using docker, for example gitlab-runner: https://stackoverflow.com/questions/73255346/how-do-you-resolve-the-gitlab-error-error-response-from-daemon-invalid-conditi

**Comment by gesellix** [comment]
Also, the binaries report an api version 1.42, which hasn't even released by Docker/Moby, yet.
This script is used for GitHub's [Windows runners](https://github.com/actions/runner-images/pull/6220), making some projects fail due to wrong assumptions.

Would it be possible to enhance the script so that with a given "desired/requested" version, it would fetch an official release, and only if no desired version is given to fetch a latest build?

**Comment by slonopotamus** [comment]
Given the current sad situation with official releases from `moby/moby`, possibly one might want to switch to a more stable source.

For example, [this](https://github.com/slonopotamus/stevedore#chocolatey).

**Comment by neersighted** [comment]
https://download.docker.com/win/static/stable/x86_64/ is available for Moby 23.0 and some versions of 20.10; and releases in moby/moby will come much faster now that we have shipped "the big one." That being said, it's unclear why nightly builds were ever used -- maybe because Docker Inc. wasn't building Windows binaries for some versions of 20.10?

Regardless, I'd invite Microsoft to come participate in moby/moby more; it's still important infrastructure as the only tool that can build Windows containers, and I hope we can continue to keep it useful and relevant for Windows users.

**Comment by jankap** [comment]
I don't understand how this is not fixed yet? [The official documentation from MS](https://learn.microsoft.com/en-us/virtualization/windowscontainers/quick-start/set-up-environment?tabs=dockerce#tabpanel_1_dockerce) directly links to the install script which installs a very old version of Docker...

**Comment by neersighted** [comment]
There is no official guidance from Docker Inc. on installing Docker CE on Windows, so that is the most likely cause. There are official binaries (with code-signing, if I'm not mistaken), but since Microsoft has maintained all the install tooling there is just a bit of a disconnect.

Still, there are plenty of options; if someone delivers a PR rewriting the installer script I imagine this can get unstuck much faster.

**Comment by slonopotamus** [comment]
You may want to try [Stevedore](https://github.com/slonopotamus/stevedore) for convenient installation of Docker Engine on both Windows Server and Client. It provides installer in an MSI form and is also available via Chocolatey and WinGet package managers.

**Comment by neersighted** [comment]
I think that if Microsoft opts for a non-Docker CE build of Moby, they're more likely to build internally (e.g. their Azure builds used by ADO/GHA) than to use something from a third party not affiliated with the project.

That being said, your project looks interesting @slonopotamus; I'd invite you to get involved upstream as an interesting downstream/packager. We could add a mention of your distribution when we get around to compiling a list, and awareness of your user-base as you grow it could be quite useful.

**Comment by slonopotamus** [comment]
> I think that if Microsoft opts for a non-Docker CE build of Moby, they're more likely to build internally (e.g. their Azure builds used by ADO/GHA) than to use something from a third party not affiliated with the project.

My point is that there's no need to wait for any moves from Microsoft in order to be able to use Docker on Windows comfortly. Moby is itself a thirdparty project, but for whatever reasons it doesn't provide installer itself (if it did, there would be no reason for Stevedore to exist). And the task of creating an installer is trivial enough so that a single person can do it in their spare time.

**Comment by neersighted** [comment]
Yes, but this issue is not about "there is no way to install Docker CE on Windows;" this issue is about the Microsoft docs and Azure DevOps/GitHub Actions runners suggesting a Microsoft-maintained script that installs stale nightly versions.

Also, Stevedore looks neat, but it's not "Docker CE," it's another distribution of similar but different components with its own lifecycle, versioning, and does not use the Docker trademark. It's a good answer as to "how do I get Moby (a Docker-compatible container engine) installed," but does not answer "how do I get Docker CE installed."

**Comment by slonopotamus** [comment]
I'm not sure what is "Docker CE". There are binaries distributed at https://download.docker.com/win/static/stable/x86_64/. Is that Docker CE?

**Comment by slonopotamus** [comment]
And yes, I agree that both docs and GitHub Actions runners using nightly builds is not a good idea WRT stability.

**Comment by neersighted** [comment]
"Docker CE" is the free, community-supported distribution of Moby provided by Docker Inc. It includes (or at least, will attempt to include; the builds are broken right now) an opinionated set of additional tools like Docker Compose, Docker Buildx, and Docker SBOM.

The Moby project is the open source upstream we use to collaborate; it is not the same as Docker CE as Docker CE is a free product/distribution produced by Docker Inc. Microsoft specifically does not provide Docker CE in GHA/Azure DevOps Ubuntu runners as they prefer to provide their own Azure builds that do not have the Docker branding and make difference decisions as to what code is included.

---

This is still something we're working to clarify as Docker Inc. and Mirantis, two of the biggest companies providing Moby distributions (Docker Desktop, Docker CE, Mirantis Container Runtime), and others can probably help outline the historical context more. I just want to make clear that the point of the Moby project was to allow for others to build on top of a neutral upstream and to split Docker Inc. products from the open source project so that it could have shared ownership and stand on its own merits.

What I'm trying to point out is that your project has its own lifecycle, versioning, and selection of components/component versions. It looks like a great way to get started with a useful Windows Containers development environment and I'd love to promote it upstream in the Moby project, but I don't think it's a good answer for this issue, which should likely decide that "Docker" is Docker CE and should come from Docker Inc., or "Docker" is a "Docker-compatible runtime" and Microsoft would rather build Moby themselves.

In short, I think this issue is a matter of who is trusted to build binaries and if "Docker in GHA" means "Docker-compatible runtime" or "Docker CE."

**Comment by slonopotamus** [comment]
> the builds are broken right now

I'm interpreting that as "such thing as Docker CE doesn't exist currently", sorry. And this is _exactly_ the reason I had to do what I did.

Anyway. I'm just offering an alternative way to get things done. There's no multi-tens-thousands dollars of enterprise support, it's just a hobby project, so there are reasons why others might prefer to use Mirantis offerings.

Peace to everyone, let's try to make the world a better place.

**Comment by neersighted** [comment]
I don't think the Mirantis offering should be preferred; it's license-encumbered and not freely distributable. Docker CE is not encumbered, the only baggage it comes with is you get Docker Inc.'s choice of plugins/additional code and the Docker trademark.

When I say "the builds are broken right now," I mean that it doesn't ship with the bundled plugins; it's "just" the CLI and daemon. Still, it's Docker CE; once Docker CE on Windows includes plugins, those would be picked up.

Or maybe to simplify: if Microsoft provides Docker CE on Windows, you'll get Docker Inc.'s opinions WRT additional components and their versions; if Microsoft does their own builds, changes Docker Inc. makes to Docker CE like said additional components being bundled will not affect GHA/Azure DevOps, or users of their install script.

---

I hope I'm not overwhelming you/discouraging you, I think your project provides a vital service. What I'm trying to explain is that I think that Microsoft is unlikely to find binaries built by you to be useful to incorporate into their own products (the crux of this issue); however a PR to Microsoft's docs plugging your project **in favor of the script**, as well as us mentioning your project upstream when we get around to documenting **how end users can get the Moby code on their system** would be quite valuable.

moby/moby is for developers, not for end users, that bit is clearly explained. The thing that we could improve upstream is documenting where end users can go to actually find a useful binary distribution of the Moby project code. Right now those include:
* Docker CE (free, available for Linux distros and Windows)
* Docker Desktop (paid, VM-isolated)
* Mirantis Container Runtime (paid, enterprise-focuesed)
* balenaEngine (free, hotrodded/a very divergent soft fork for IOT)
* Rancher Kubernetes Engine (free, a "best effort" / "compatibility layer" feature)
* Stevedore (free, minimal distribution for Windows Containers on Windows)
* GHA/Azure Devops builds (only part of the MS product, Linux-only right now)

The versioning/lifecycle for these different distributions can end up being quite different; currently we're planning (as Mirantis and Microsoft) to start maintaining longer-lived "stable" branches upstream, so MCR and the GHA/Azure DevOps distributions are likely to stay on an older release branch while Docker CE becomes more bleeding edge, as one example.

Technically there are others like Huawei Cloud Container Engine (a RedHat-style compilation of backports to a very old version), Unraid (which builds their own as best I can tell, instead of consuming Docker CE), Synology, QNAP, etc., but those are so inflexible/disconnected from upstream that I'm not sure directing users there is useful.

**Comment by slonopotamus** [comment]
> Microsoft is unlikely to find binaries built by you to be useful to incorporate into their own products

You possibly missed something. Stevedore is not building its own Docker/Moby binaries. It just bundles things from https://download.docker.com/win/static/stable/x86_64/ into an MSI installer. So, from a technical POV, the "Docker CE"ness of thing you get from it is equal to what you get from `install-docker-ce.ps1`.

And no, I'm not insisting on Stevedore to become a way officially recommended by Microsoft. As I already said, I'm just plugging a hole between those binaries and end-user machine.


----

As you can see, I even [verify using SHA256](https://github.com/slonopotamus/stevedore/blob/main/build.rs#L13-L15) that what is bundled is _exactly_ what Docker, Inc has built.

**Comment by neersighted** [comment]
Ah, I did miss that it was a redistribution and not a new distribution. In that case you probably can (and should) explain that in the README so users understand they're getting the binaries directly from Docker Inc., and that the decisions that Docker Inc. makes WRT their distribution will be reflected in Stevedore.

`install-docker-ce.ps1` incidentally never installed Docker CE, but actually installed nightly binaries that used to be built by Moby Project CI, leading to the confusion of product identity in general/in this issue :laughing:

---

In any case, thanks for introducing Stevedore and I'll try to surface it to the other maintainers/contributors; I think there's certainly a (unrelated to this issue) dialogue to be had about how we can improve the install experience both as part of Moby, and for Docker Inc. -- your project could very well be part of that conversation if you want it to be.

**Comment by slonopotamus** [comment]
> a PR to Microsoft's docs plugging your project in favor of the script, as well as us mentioning your project upstream when we get around to documenting how end users can get the Moby code on their system would be quite valuable.

Okay, I'll try creating one (why not) :)

**Comment by slonopotamus** [comment]
> the decisions that Docker Inc. makes WRT their distribution will be reflected in Stevedore.

Absolutely 100%. Stevedore doesn't have neither will nor manpower to deviate from upstream. It's scope is very specific and limited: provide a way to install what Docker, Inc has built onto user machines (both Windows Server and Windows Client). And the day when Docker, Inc provides its own way to do the same task, Stevedore will likely to be shut down because it's goal will now be achieved even without its existensce.

**Comment by neersighted** [comment]
I suppose it is a little more than Docker CE as you also include the credential helper and compose, which are not part of the base Docker CE right now; however it's probably best to simply promote the bit in the README where you state that you compile existing binaries, and make the provenance clear:
* Docker CE (currently engine + CLI)
* Docker Compose (v2, plugin)
* Credential helpers as used by Desktop

---

I think we're now firmly off topic for this issue; if you'd like to chat more about this (and potentially where it will be seen by the other Moby maintianers), how about we take it to the [Docker Community Slack](https://dockr.ly/comm-slack)?

**Comment by brasmith-ms** [comment]
Hey folks, apologies was out for a bit and missed the updates on this thread. 

Firstly, the purpose of this script is just to be a useful utility to get up and running quickly. When I first created this I thought the master.dockerproject.org site would be kept up to date, but https://github.com/moby/moby/issues/45081 explains this is not the case. 

I will update this script to utilize the download.dockerproject.org link instead with a version check for the latest. 

Secondly, this script is open-source and I highly encourage contributions to it to add things the community finds useful. The MS docs point to it as a way to get the bare minimum installed to run containers with docker with just basic powershell functionality. I am more than happy to add separate community projects as alternatives on the docs, as we did for the [containerd installer](https://learn.microsoft.com/en-us/virtualization/windowscontainers/quick-start/set-up-environment?tabs=containerd). So @slonopotamus if you have a docs PR link please send that to me!

And finally, I'm always happy to chat about things like this, I check the Kubernetes slack somewhat regularly so feel free to [reach out to me](kubernetes.slack.com) there (search for **Brandon Smith (MS)**).

**Comment by mthalman** [comment]
> Firstly, the purpose of this script is just to be a useful utility to get up and running quickly. 

Given that this script is being [used to configure GitHub Action runner images](https://github.com/actions/runner-images/blob/c22eb26949bfb73894a6020e0e77886dbe3af52a/images/win/scripts/Installers/Install-Docker.ps1#L28), it probably needs more attention than what you intended it to be.

**Comment by brasmith-ms** [comment]
Folks, please check out #331 and add any suggestions.
